### ISSUE TITLE
Predict proba with changing seed

### ISSUE THREAD

**Issue by amirsani** [awslabs/autogluon#425]
Is there a way to run the model or train the model with different random seeds?

Thanks

**Comment by Innixma** [comment]
Do you have a particular task in mind (Tabular, Image, etc.)? Regarding Tabular, this is not added at present but could be if you believe there is value in it.

**Comment by amirsani** [comment]
There's value in that it provides an avenue for generating bootstrap confidence intervals in the case of prediction and robustness in the case of training. Note that early stopping has a lot of value in the case of training.

Thanks for the great package and putting the time into this.

**Comment by Innixma** [comment]
Just to confirm, are you referring to this functionality for Tabular? Also, how does early stopping relate to this exactly?

**Comment by amirsani** [comment]
Yes, tabular

As you estimate the performance over seeds, you can use an online estimator to see if the mean is converging to a value that is better than the current performance threshold. You can also use forecasts to see if the rate of improvement in the estimated performance is likely to exceed the stopping threshold. You can further accelerate this estimate by bootstrapping.

**Comment by amirsani** [comment]
How can I set the seed of a model that's been learned with AutoGluon?

**Comment by Innixma** [comment]
Please see PR #429 for the addition of the random_seed parameter to the task.fit() call.

Regarding estimating performance, we have an optimized version of this if you are using bagging which is repeated bagged functionality. Using it will give you both better predictions and more accurate prediction estimates. It is enabled automatically if you specify time_limits and set auto_stack=True, or by specifying num_bagging_folds and num_bagging_sets. This will be far more efficient and produce better results than bootstrap aggregation with different random seeds as an outer loop to AutoGluon.

See also our tutorial on [maximizing model performance](https://autogluon.mxnet.io/tutorials/tabular_prediction/tabular-indepth.html#maximizing-predictive-performance) in Tabular.

Regarding calculating the confidence intervals inside of AutoGluon without an outer loop, this would require getting the predictions of the individual models within a bag rather than the averaged predictions of the bagged models. This is something users do not have access to from the user-facing API at present, partly because of the ease at which it could cause OOM errors on multiclass datasets.

Regarding 

> How can I set the seed of a model that's been learned with AutoGluon?

Could you explain your intent? Do you mean setting the seed prior to fitting the model (#429 implements this)? There is no randomness associated with an already trained model, and thus no seed to set after training.

**Comment by amirsani** [comment]
Thank you for the quick turnaround and apologies for my poor explanation.

I would like to take the top model output by AutoGluon, let's call it model, and refit that specific model after training to the full data with different seeds. I believe it might look like model.refit_full(seed), with the option of applying it to new data, so maybe something like, 
``
seeds = range(100)
models = [model.fit(X,y,seed=i) for i in seeds]
Predictions = [model.predict_proba(X_out_of_sample) for model in models]
``

**Comment by Innixma** [comment]
Ah, I see what you mean now!

To support this I need to enable the following functionality:

1. Enable user to create an unfit copy of a model.
2. Enable user to alter the hyperparameters of the unfit model copy to update whatever seed parameter the model might have.
3. Model training after the task.fit() call (predictor.fit_continue() or something similar).

#1 is already available internally, I would just have to expose it in the predictor API.
#2 is not difficult, but an API may have to be created for it to avoid user confusion when accessing internal model variables.
#3 is trickier because I need to make it flexible while not necessarily giving all the options that the internal trainer API has (because that would be severely confusing to users).

The reason I am hesitant to simply add the seed parameter alone as you show (model.refit_full(X, y, seed) would work) is because while that will satisfy your use-case, countless others will not be satisfied (perhaps a user wants to change model learning rate, or alter the base_model composition, etc.)

Let me know if the steps I mentioned satisfy your use-case and I will look towards prioritizing them.
